cleverhans/attacks_eager.py

```diff
@@ -0,0 +1,260 @@
+            elif ord == 2:
+                # avoid_zero_div must go inside sqrt to avoid a divide by zero
+                # in the gradient through this operation
+                norm = tf.sqrt(tf.maximum(avoid_zero_div,
+                                          tf.reduce_sum(tf.square(eta),
+                                                        reduc_ind,
+                                                        keep_dims=True)))
+            # We must *clip* to within the norm ball, not *normalize* onto the
+            # surface of the ball
+            factor = tf.minimum(1., eps / norm)
+            eta = eta * factor
+        return eta
+
+
+class FastGradientMethod(Attack):
+    """
+    This attack was originally implemented by Goodfellow et al. (2015) with the
+    infinity norm (and is known as the "Fast Gradient Sign Method"). This
+    implementation extends the attack to other norms, and is therefore called
+    the Fast Gradient Method.
+    Paper link: https://arxiv.org/abs/1412.6572
+    """
+    def __init__(self, model, dtypestr='float32'):
+        """
+        Create a FastGradientMethod instance.
+        Note: the model parameter should be an instance of the
+        cleverhans.model.Model abstraction provided by CleverHans.
+        """
+        if not isinstance(model, Model_Eager):
+            model = CallableModelWrapper(model, 'probs')
+
+        super(FastGradientMethod, self).__init__(model, dtypestr)
+
+    def generate(self, x, **kwargs):
+        """
+        Generates the adversarial sample for the given input.
+
+        :param x: The model's inputs.
+        :param eps: (optional float) attack step size (input variation)
+        :param ord: (optional) Order of the norm (mimics NumPy).
+                    Possible values: np.inf, 1 or 2.
+        :param y: (optional) A tf variable` with the model labels. Only provide
+                  this parameter if you'd like to use true labels when crafting
+                  adversarial samples. Otherwise, model predictions are used as
+                  labels to avoid the "label leaking" effect (explained in this
+                  paper: https://arxiv.org/abs/1611.01236). Default is None.
+                  Labels should be one-hot-encoded.
+        :param y_target: (optional) A tf variable` with the labels to target.
+                            Leave y_target=None if y is also set.
+                            Labels should be one-hot-encoded.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        # Parse and save attack-specific parameters
+        assert self.parse_params(**kwargs)
+        labels, nb_classes = self.get_or_guess_labels(x, kwargs)
+        return self.fgm(x, labels=labels, targeted=(self.y_target is not None))
+
+    def parse_params(self, eps=0.3, ord=np.inf, y=None, y_target=None,
```

Comment:
What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

Reply:
Same as before, lets discuss in the previous comment.